backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java

```diff
@@ -0,0 +1,82 @@
+package com.redhat.cloud.notifications.processors.email;
+
+import com.redhat.cloud.notifications.db.EndpointEmailSubscriptionResources;
+import com.redhat.cloud.notifications.ingress.Action;
+import com.redhat.cloud.notifications.models.Application;
+import com.redhat.cloud.notifications.models.EmailSubscription;
+import com.redhat.cloud.notifications.models.Endpoint;
+import com.redhat.cloud.notifications.models.EndpointType;
+import com.redhat.cloud.notifications.models.Notification;
+import com.redhat.cloud.notifications.models.NotificationHistory;
+import com.redhat.cloud.notifications.processors.webhooks.WebhookTypeProcessor;
+import io.quarkus.test.junit.QuarkusTest;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import javax.inject.Inject;
+import java.time.LocalDateTime;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static com.redhat.cloud.notifications.models.EmailSubscriptionType.*;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@QuarkusTest
+class EmailSubscriptionTypeProcessorTest {
+
+    @Inject
+    EmailSubscriptionTypeProcessor testee;
+
+    @Test
+    void shouldNotProcessWhenEndpointsAreNull() {
+        final Action action = new Action("someBundle", "someApplication", "someEventType", LocalDateTime.now(), "someAccountId", Map.of(), List.of());
+        final Multi<NotificationHistory> process = testee.process(action, null);
+
+        assertEquals(0, process.collect().asList().await().indefinitely().size());
+    }
+
+    @Test
+    void shouldNotProcessWhenEndpointsAreEmpty() {
+        final Action action = new Action("someBundle", "someApplication", "someEventType", LocalDateTime.now(), "someAccountId", Map.of(), List.of());
+        final Multi<NotificationHistory> process = testee.process(action, List.of());
+
+        assertEquals(0, process.collect().asList().await().indefinitely().size());
+    }
+
+    @Test
+    @Disabled
+    void shouldProcess() {
+        final Action action = new Action("someBundle", "someApplication", "someEventType", LocalDateTime.now(), "someAccountId", Map.of(), List.of());
+
+        Endpoint endpoint = new Endpoint();
+        endpoint.setType(EndpointType.EMAIL_SUBSCRIPTION);
+        endpoint.setName("someName");
+        endpoint.setEnabled(true);
+        endpoint.setDescription("someDescription");
+        endpoint.setAccountId("someAccountId");
+
+        testee.process(action, List.of(endpoint));
+    }
+
+    @Test
+    void shouldSendEmail() {
```

Comment:
`EmailSubscriptionTypeProcessor#consumeEmailAggregations` should be tested using [SmallRye Reactive Messaging testing features](https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/testing/testing.html). There is an example available in [EventConsumerTest](https://github.com/RedHatInsights/notifications-backend/blob/master/backend/src/test/java/com/redhat/cloud/notifications/events/EventConsumerTest.java#L42).

The idea is to replace the incoming Kafka `aggregation` channel with a mocked channel `InMemoryConnector.source("aggregation")` and then send a payload throught that channel. It will trigger a reactive messaging processing in `EmailSubscriptionTypeProcessor#consumeEmailAggregations` exactly like what would happen on prod. Then you should be able to run the Mockito `verify`.

Reply:
I am not able to test it. Debugging doesn't work, changing the method to:     

@Incoming("aggregation")
    public Uni<Void> consumeEmailAggregations(String aggregationsJson) throws JsonProcessingException {throw new IllegalArgumentException("");}

Also doesn't result in exceptions in tests. I lost so many hours and days with this stuff and don't want to lose more. Testing single methods which deal with Unis and Multis doesn't work and testing this is also not fun. I will stop it here and mark it as resolved.

Reply:
I can try fixing this if you want.

Reply:
I pushed a test for the Kafka messages processing. This only works for successfull test cases.

Testing a failure because of an invalid payload will require another kind of test based on [this](https://smallrye.io/smallrye-mutiny/guides/testing).

Reply:
Thanks :) I actually wanted to test more specific things, but maybe this is enough.